File: app/src/main/java/org/jay/launchstarterexample/MyApplication.java
package org.jay.launchstarterexample;

import android.app.Application;
import android.content.Context;
import android.os.StrictMode;
import androidx.multidex.MultiDex;
import com.tencent.mmkv.MMKV;
import org.jay.launchstarter.TaskDispatcher;
import org.jay.launchstarterexample.bean.NewsItem;
import org.jay.launchstarterexample.tasks.*;
import org.jay.launchstarterexample.utils.LaunchTimer;

public class MyApplication extends Application {

    private String mDeviceId;
    private static Application mApplication;
    private boolean DEV_MODE = true;

    @Override
    protected void attachBaseContext(Context base) {
        super.attachBaseContext(base);
        MultiDex.install(this);
    }

    @Override
    public void onCreate() {
        super.onCreate();
        mApplication = this;

        TaskDispatcher.init(this);
        TaskDispatcher dispatcher = TaskDispatcher.createInstance();
        dispatcher
                .addTask(new InitAMapTask())
                .addTask(new InitStethoTask())
                .addTask(new InitWeexTask())
                .addTask(new InitBuglyTask())
                .addTask(new InitFrescoTask())
                .addTask(new InitJPushTask())
                .addTask(new InitUmengTask())
                .addTask(new GetDeviceIdTask())
                .start();

        dispatcher.await();
        initStrictMode();
    }

    private void initStrictMode() {
        if (DEV_MODE) {
            StrictMode.setThreadPolicy(new StrictMode.ThreadPolicy.Builder()
                    .detectCustomSlowCalls() //API等级11，使用StrictMode.noteSlowCode
                    .detectDiskReads()
                    .detectDiskWrites()
                    .detectNetwork()// or .detectAll() for all detectable problems
                    .penaltyLog() //在Logcat 中打印违规异常信息
                    .build());
            StrictMode.setVmPolicy(new StrictMode.VmPolicy.Builder()
                    .detectLeakedSqlLiteObjects()
                    .setClassInstanceLimit(NewsItem.class, 1)
                    .detectLeakedClosableObjects() //API等级11
                    .penaltyLog()
                    .build());
        }
    }

    public void setDeviceId(String deviceId) {
        this.mDeviceId = deviceId;
    }

    public String getDeviceId() {
        return mDeviceId;
    }

    public static Application getApplication() {
        return mApplication;
    }

}

File: app/src/main/java/org/jay/appstarterexample/MyApplication.kt
package org.jay.appstarterexample

import android.app.Application
import android.content.Context
import android.os.StrictMode
import androidx.multidex.MultiDex
import org.jay.appstarterexample.MyApplication
import org.jay.appstarter.TaskDispatcher
import org.jay.appstarterexample.tasks.InitAMapTask
import org.jay.appstarterexample.tasks.InitStethoTask
import org.jay.appstarterexample.tasks.InitWeexTask
import org.jay.appstarterexample.tasks.InitBuglyTask
import org.jay.appstarterexample.tasks.InitFrescoTask
import org.jay.appstarterexample.tasks.InitJPushTask
import org.jay.appstarterexample.tasks.InitUmengTask
import org.jay.appstarterexample.tasks.GetDeviceIdTask
import android.os.StrictMode.ThreadPolicy
import android.os.StrictMode.VmPolicy
import org.jay.appstarterexample.tasks.delayinittask.DelayInitTaskA
import org.jay.appstarterexample.tasks.delayinittask.DelayInitTaskB

/**
 * 是否为开发模式
 */
private val DEV_MODE = BuildConfig.DEBUG

class MyApplication : Application() {

    /**
     * 设备 ID
     */
    var deviceId: String? = null

    override fun attachBaseContext(base: Context) {
        super.attachBaseContext(base)
        MultiDex.install(this)
    }

    override fun onCreate() {
        super.onCreate()
        application = this

        // 初始化任务分发器
        TaskDispatcher.init(this)

        // 添加并开始执行初始化任务
        val dispatcher = TaskDispatcher.createInstance()
        dispatcher
            .addTask(InitAMapTask())    // 高德 SDK 初始化任务
            .addTask(InitStethoTask())  // Stetho 初始化任务
            .addTask(InitWeexTask())    // weex 初始化任务
            .addTask(InitBuglyTask())   // Bugly 初始化任务
            .addTask(InitFrescoTask())  // Frescode 初始化任务
            .addTask(InitJPushTask())   // 极光推送 SDK 初始化任务
            .addTask(InitUmengTask())   // 友盟 SDK 初始化任务
            .addTask(GetDeviceIdTask()) // 获取设备 ID 初始化任务
            .addTask(DelayInitTaskA())  // 延迟初始化任务 A
            .addTask(DelayInitTaskB())  // 延迟初始化任务 B
            .start()
        dispatcher.await()
        initStrictMode()
    }

    private fun initStrictMode() {
        if (DEV_MODE) {
            StrictMode.setThreadPolicy(
                ThreadPolicy.Builder()
                    .detectCustomSlowCalls() //API等级11，使用StrictMode.noteSlowCode
                    .detectDiskReads()
                    .detectDiskWrites()
                    .detectNetwork() // or .detectAll() for all detectable problems
                    .penaltyLog() //在Logcat 中打印违规异常信息
                    .build()
            )
            StrictMode.setVmPolicy(
                VmPolicy.Builder()
                    .detectLeakedSqlLiteObjects()
                    .detectLeakedClosableObjects() //API等级11
                    .penaltyLog()
                    .build()
            )
        }
    }

    companion object {
        var application: Application? = null
            private set
    }
}